### DISCUSSION TITLE
[Week 16] 作業第四題，在 global 底下是任何變數名稱都會先被宣告嗎?

### DISCUSSION THREAD

**Discussion by i958047527992** [Lidemy/forum#188] [程式導師實驗計畫] [answered]
```
const obj = {
  value: 1,
  hello: function() {
    console.log(this.value)
  },
  inner: {
    value: 2,
    hello: function() {
      console.log(this.value)
    }
  }
}
  
const obj2 = obj.inner
const hello = obj.inner.hello
obj.inner.hello() // ??
obj2.hello() // ??
hello() // ??
```
對於第三個的 hello() 來說，裡面的 this 會變成 global，所以會輸出的是 global.value 的值，為 undefined。
![image](https://user-images.githubusercontent.com/9329023/130041070-99d19a44-1eab-48a6-8828-a1c3f8dd88ad.png)

那這個 value 到底是在何時被宣告的?
以 node js 來說，對於沒有宣告的變數，輸出的話會出現 `Uncaught ReferenceError: <變數名稱> is not defined`，只有在宣告之後才會變成 undefined。
![image](https://user-images.githubusercontent.com/9329023/130041399-e0e3b059-af1c-43b5-b80f-78c720dfb1ea.png)

但 global 底下似乎不管任何變數名稱，都會得到 undefined，到底是哪裡被宣告的?
![image](https://user-images.githubusercontent.com/9329023/130041192-44ca280f-c04b-4f95-a750-a90a60ca67d4.png)

**Comment by aszx87410** [comment, marked as answer]
無論是 `typeof global` 或是 `typeof window`，結果都是 object
所以你存取 global.value 就是在讀取一個物件不存在的 key，所以就是 undefined，跟宣告沒關係

就像 const obj = {}; console.log(obj.value) // undefined

**Reply by i958047527992** [reply]
原來如此，感謝胡立解答
